{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene matrix recipe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import statements\n",
    "import os\n",
    "import yaml\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path settings\n",
    "project_dir = os.path.dirname(os.path.abspath('.'))\n",
    "conf_dir = os.path.join(project_dir, 'conf')\n",
    "path_conf_path = os.path.join(conf_dir, 'filepaths.yaml')\n",
    "gene_list_conf_path = os.path.join(conf_dir, 'gene_list_names.yaml')\n",
    "gene_type_conf_path = os.path.join(conf_dir, 'gene_biotype.yaml')\n",
    "\n",
    "with open(path_conf_path) as path_conf_file:\n",
    "    path_dict = yaml.safe_load(path_conf_file)\n",
    "    \n",
    "with open(gene_list_conf_path) as gene_list_conf_file:\n",
    "    gene_list_name_dict = yaml.safe_load(gene_list_conf_file)\n",
    "\n",
    "gencode_path = os.path.join(project_dir, path_dict['GENCODE'])\n",
    "prev_gene_mat_path = os.path.join(project_dir, path_dict['An2018'])\n",
    "hgnc_path = os.path.join(project_dir, path_dict['HGNC'])\n",
    "alt_gene_list_path = os.path.join(project_dir, path_dict['An2018_ALT_GENE'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "gencode_df = pd.read_table(gencode_path, compression='gzip', comment='#', names=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'])  \n",
    "gencode_gene_df = gencode_df[gencode_df['feature'] == 'gene']  # List up only genes\n",
    "gencode_tx_df = gencode_df[gencode_df['feature'] == 'transcript']  # List up only transcripts\n",
    "hgnc_df = pd.read_table(hgnc_path, usecols=['hgnc_id', 'symbol', 'alias_symbol', 'prev_symbol', 'ensembl_gene_id', 'ucsc_id', 'refseq_accession'])\n",
    "prev_gene_mat_df = pd.read_excel(prev_gene_mat_path, sheet_name='8-1 Genesets')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to parse strings of the 'attribute' field in the GENCODE.\n",
    "def parse_attr_str(attr_str):\n",
    "    attrs = attr_str.split(';')\n",
    "    attr_dict = {}\n",
    "    \n",
    "    for attr in attrs:\n",
    "        key, value = attr.split('=')\n",
    "        attr_dict[key] = value\n",
    "        \n",
    "    return attr_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse values of the 'attribute' field in the GENCODE\n",
    "gene_to_attr_dict = {}  # Key: GeneID, Value: A dictionary for the information in the 'attribute' columns\n",
    "gene_name_to_ids = defaultdict(list)  # Key: Gene name, Value: The list of gene IDs\n",
    "hgnc_to_ids = defaultdict(list)  # Key: HGNC ID, Value: The list of gene IDs\n",
    "\n",
    "for attr_str in gencode_gene_df['attribute'].values:\n",
    "    attr_dict = parse_attr_str(attr_str)\n",
    "    \n",
    "    if not attr_dict['ID'].endswith('Y'):  # Ignore pseudoautosome region (PAR_Y)\n",
    "        gene_id = attr_dict['ID'].split('.')[0]\n",
    "        gene_to_attr_dict[gene_id] = attr_dict\n",
    "        gene_name_to_ids[attr_dict['gene_name']].append(gene_id)\n",
    "        hgnc_id = attr_dict.get('hgnc_id')\n",
    "        \n",
    "        if hgnc_id is not None:\n",
    "            hgnc_to_ids[hgnc_id].append(gene_id)\n",
    "\n",
    "tx_to_attr_dict = {}  # Key: GeneID, Value: A dictionary for the information in the 'attribute' columns\n",
    "\n",
    "for attr_str in gencode_tx_df['attribute'].values:\n",
    "    attr_dict = parse_attr_str(attr_str)\n",
    "    \n",
    "    if not attr_dict['ID'].endswith('Y'):  # Ignore pseudoautosome region (PAR_Y)\n",
    "        tx_id = attr_dict['ID'].split('.')[0]\n",
    "        tx_to_attr_dict[tx_id] = attr_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the HGNC file and make a dictionary which key and value are a gene symbol and its ensembl gene ID, respectively.\n",
    "hgnc_df_col_idx = {column: i for i, column in enumerate(hgnc_df.columns.values)}\n",
    "hgnc_df_val = hgnc_df.values\n",
    "\n",
    "symbol_to_gene_ids = {}\n",
    "alias_to_gene_ids = {}\n",
    "prev_symbol_to_gene_ids = {}\n",
    "\n",
    "for hgnc_entry in hgnc_df_val:\n",
    "    gene_symbol = hgnc_entry[hgnc_df_col_idx['symbol']]\n",
    "    alias_symbol_str = hgnc_entry[hgnc_df_col_idx['alias_symbol']]\n",
    "    prev_symbol_str = hgnc_entry[hgnc_df_col_idx['prev_symbol']]\n",
    "    ensembl_gene_id = hgnc_entry[hgnc_df_col_idx['ensembl_gene_id']]\n",
    "    \n",
    "    if ensembl_gene_id is np.nan or gene_to_attr_dict.get(ensembl_gene_id) is None:\n",
    "        gene_ids = None\n",
    "    else:\n",
    "        gencode_symbol = gene_to_attr_dict[ensembl_gene_id]['gene_name']\n",
    "        gene_ids = gene_name_to_ids.get(gencode_symbol)  # From GENCODE v33\n",
    "    \n",
    "    if gene_ids is not None: \n",
    "        symbol_to_gene_ids[gene_symbol] = gene_ids\n",
    "\n",
    "        if alias_symbol_str is not np.nan:\n",
    "            alias_symbols = alias_symbol_str.split('|')\n",
    "\n",
    "            for alias_symbol in alias_symbols:\n",
    "                alias_to_gene_ids[alias_symbol] = gene_ids\n",
    "\n",
    "        if prev_symbol_str is not np.nan:\n",
    "            prev_symbols = prev_symbol_str.split('|')\n",
    "\n",
    "            for prev_symbol in prev_symbols:\n",
    "                prev_symbol_to_gene_ids[prev_symbol] = gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to find a gene ID from a gene symbol of the HGNC and the GENCODE v33\n",
    "def find_gene_ids(gene_symbol):\n",
    "    # Priority: \n",
    "    # 1. GENCODE v33\n",
    "    # 2. HGNC: symbol -> alias -> previous symbol\n",
    "    # If the gene IDs cannot be found, return None\n",
    "    \n",
    "    # GENCODE v33\n",
    "    gene_ids = gene_name_to_ids.get(gene_symbol)\n",
    "    \n",
    "    # HGNC\n",
    "    if gene_ids is None:\n",
    "        gene_ids = symbol_to_gene_ids.get(gene_symbol)\n",
    "\n",
    "        if gene_ids is None:\n",
    "            gene_ids = alias_to_gene_ids.get(gene_symbol)\n",
    "\n",
    "            if gene_ids is None:\n",
    "                gene_ids = prev_symbol_to_gene_ids.get(gene_symbol)\n",
    "\n",
    "    return gene_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Make a list of alternative gene names and IDs \n",
    "The purpose of this step is to find genes *deprecated in the GENCODE v33* from the *8-1 Genesets* sheet of the **Supplementary Table S8** in An *et al.*, *Science*, 2018 (This is a **previous gene matrix**).\n",
    "\n",
    "*Note: If you alreday have the file of {alt_gene_list_path}, skip this step.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No. deprecated genes: 482\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Genes                             AC000041.2AC002310.4AC002384.2AC002465.1AC0024...\n",
       "EnsemblGeneId                     ENSG00000274457.1ENSG00000260869.1ENSG00000283...\n",
       "ASD associated genes (FDR≤0.1)                                                    0\n",
       "ASD associated genes (FDR≤0.3)                                                    0\n",
       "Midfetal co-expression genes                                                      0\n",
       "Brain expressed genes                                                            11\n",
       "Constrained genes (pLI ≥ 0.9)                                                     0\n",
       "Postsynaptic density genes                                                        1\n",
       "Developmental delay genes                                                         2\n",
       "CHD8 targets                                                                      0\n",
       "FMRP targets                                                                      0\n",
       "Protein Coding                                                                   59\n",
       "mis3_pro                                                                          0\n",
       "mis3_sib                                                                          0\n",
       "Prom_pro                                                                         25\n",
       "Prom_sib                                                                         21\n",
       "Prom_ActiveTSS_pro                                                                8\n",
       "Prom_ActiveTSS_sib                                                               13\n",
       "Prom_ConservedLoci_pro                                                            5\n",
       "Prom_ConservedLoci_sib                                                            1\n",
       "dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate No. of not available genes in the previous gene matrix\n",
    "gene_id_set = set(gene_to_attr_dict.keys())  # From the GENCODE v33\n",
    "prev_gene_ids = np.vectorize(lambda x: x.split('.')[0])(prev_gene_mat_df['EnsemblGeneId'].values)\n",
    "is_depr_gene = np.vectorize(lambda x: x not in gene_id_set)(prev_gene_ids)\n",
    "print(f'No. deprecated genes: {sum(is_depr_gene)}')\n",
    "depr_gene_mat_df = prev_gene_mat_df[is_depr_gene]\n",
    "depr_gene_mat_df.sum(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The list of deprecated genes with their alternative names and IDs\n",
    "# The genes in the gene lists except GENCODE biotypes and results of An et al., 2018 were considered.\n",
    "# Columns: gene, gene_id, alt_gene, alt_gene_id\n",
    "# Ref: https://www.genecards.org, http://asia.ensembl.org\n",
    "depr_gene_entries = [\n",
    "    ['AKAP2', 'ENSG00000241978.9', 'PALM2AKAP2', 'ENSG00000157654.19'],\n",
    "    ['BTBD8', 'ENSG00000284413.2', 'BTBD8', 'ENSG00000189195.13'],\n",
    "    ['C2orf48', 'ENSG00000163009.8', 'RRM2', 'ENSG00000171848.15'],\n",
    "    ['C3orf36', 'ENSG00000221972.3', 'SLCO2A1', 'ENSG00000174640.15'],\n",
    "    ['C8orf44', 'ENSG00000213865.7', 'SGK3', 'ENSG00000104205.15'],\n",
    "    ['C9orf47', 'ENSG00000186354.10', 'S1PR3', 'ENSG00000213694.5'],\n",
    "    ['HIST1H3B', 'ENSG00000274267.1', 'H3C2', 'ENSG00000286522.1'],\n",
    "    ['HIST1H3C', 'ENSG00000278272.1', 'H3C3', 'ENSG00000287080.1'],\n",
    "    ['SCO2', 'ENSG00000130489.14', 'SCO2', 'ENSG00000284194.2'],\n",
    "    ['TBCE', 'ENSG00000116957.12', 'TBCE', 'ENSG00000284770.2'],\n",
    "    ['TBCE', 'ENSG00000116957.12', 'TBCE', 'ENSG00000285053.1'],\n",
    "    ['TMEM133', 'ENSG00000170647.3', 'ARHGAP42', 'ENSG00000165895.19'],\n",
    "    ['PAML2', 'ENSG00000243444.7', 'PALM2AKAP2', 'ENSG00000157654.19'], \n",
    "]\n",
    "\n",
    "# Write the entries\n",
    "with open(alt_gene_list_path, 'w') as alt_gene_list_file:\n",
    "    print('gene', 'gene_id', 'alt_gene', 'alt_gene_id', sep='\\t', file=alt_gene_list_file)\n",
    "    \n",
    "    for gene_entry in depr_gene_entries:\n",
    "        print(*gene_entry, sep='\\t', file=alt_gene_list_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Make the previous gene matrix compatible with GENCODE v33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>FMRP_targets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  FMRP_targets  \n",
       "ENSG00000223972    0    0             0             0  \n",
       "ENSG00000227232    0    0             0             0  \n",
       "ENSG00000278267    0    0             0             0  \n",
       "ENSG00000243485    0    0             0             0  \n",
       "ENSG00000284332    0    0             0             0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the list of deprecated genes from the previous gene matrix \n",
    "alt_gene_dict = defaultdict(list)\n",
    "\n",
    "with open(alt_gene_list_path) as alt_gene_list_file:\n",
    "    alt_gene_list_file.readline()  # Read the header\n",
    "    \n",
    "    for line in alt_gene_list_file:\n",
    "        fields = line.strip().split('\\t')\n",
    "        prev_gene_id = fields[1].split('.')[0]\n",
    "        alt_gene_id = fields[3].split('.')[0]\n",
    "        alt_gene_dict[prev_gene_id].append(alt_gene_id)\n",
    "\n",
    "# Remove genes in pseudoautosome regions (PAR_Y)\n",
    "prev_gene_mat_df = prev_gene_mat_df[np.vectorize(lambda x: not x.endswith('_PAR_Y'))(prev_gene_mat_df['EnsemblGeneId'].values)]\n",
    "\n",
    "# Values from the previous gene matrix\n",
    "prev_gene_mat_values = prev_gene_mat_df.values\n",
    "prev_gene_mat_cols = prev_gene_mat_df.columns.values\n",
    "col_idx_dict = {colname: i for i, colname in enumerate(prev_gene_mat_cols)}\n",
    "\n",
    "# Get values from the previous gene matrix and store as dictionary\n",
    "prev_gene_mat_dict = {}  # Key: gene ID, Value: a dictionary (gene list name -> its value (0 or 1))\n",
    "\n",
    "for prev_gene_vals in prev_gene_mat_values:\n",
    "    gene_val_dict = {}\n",
    "    \n",
    "    # Get the values for each gene list column\n",
    "    for colname in prev_gene_mat_cols:\n",
    "        gene_list_name = gene_list_name_dict.get(colname)\n",
    "        \n",
    "        if gene_list_name is not None:  # If None, this column is not for gene lists.\n",
    "            gene_val_dict[gene_list_name] = prev_gene_vals[col_idx_dict[colname]]\n",
    "    \n",
    "    prev_gene_id = prev_gene_vals[col_idx_dict['EnsemblGeneId']].split('.')[0]\n",
    "    alt_gene_ids = alt_gene_dict.get(prev_gene_id, [prev_gene_id])  # Change the gene ID if it was deprecated\n",
    "    \n",
    "    # Make up the dictionary for the gene matrix\n",
    "    for gene_id in alt_gene_ids:\n",
    "        same_gene_dict = prev_gene_mat_dict.get(gene_id)\n",
    "        \n",
    "        # Merge values of duplicated genes using logical OR operation\n",
    "        if same_gene_dict is not None:\n",
    "            for gene_list_name in gene_val_dict:\n",
    "                gene_val_dict[gene_list_name] |= same_gene_dict[gene_list_name]\n",
    "    \n",
    "        prev_gene_mat_dict[gene_id] = dict(gene_val_dict)  # To copy deeply\n",
    "\n",
    "# Make a dictionary for the new gene matrix\n",
    "new_gene_mat_dict = {}\n",
    "\n",
    "for gene_id in gene_to_attr_dict:\n",
    "    gene_val_dict = prev_gene_mat_dict.get(gene_id, dict())\n",
    "    gene_val_dict['gene_id'] = gene_to_attr_dict[gene_id]['ID']\n",
    "    gene_val_dict['gene_name'] = gene_to_attr_dict[gene_id]['gene_name']        \n",
    "    new_gene_mat_dict[gene_id] = gene_val_dict\n",
    "\n",
    "# Make a DataFrame for the new gene matrix\n",
    "gene_mat_df = pd.DataFrame.from_dict(new_gene_mat_dict, orient='index')\n",
    "gene_mat_cols = list(gene_mat_df.columns.values)\n",
    "gene_mat_cols = gene_mat_cols[-2:] + gene_mat_cols[:-2]\n",
    "gene_mat_df = gene_mat_df[gene_mat_cols]\n",
    "gene_mat_df.fillna(0, inplace=True)\n",
    "gene_mat_df = gene_mat_df.astype({gene_list_col: 'int64' for gene_list_col in gene_mat_cols[2:]})\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Add GENCODE biotypes as gene list columns into the gene matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>FMRP_targets</th>\n",
       "      <th>Protein_Coding</th>\n",
       "      <th>Long_ncRNA</th>\n",
       "      <th>Small_ncRNA</th>\n",
       "      <th>Pseudogene</th>\n",
       "      <th>IG_TR_Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  FMRP_targets  Protein_Coding  \\\n",
       "ENSG00000223972    0    0             0             0               0   \n",
       "ENSG00000227232    0    0             0             0               0   \n",
       "ENSG00000278267    0    0             0             0               0   \n",
       "ENSG00000243485    0    0             0             0               0   \n",
       "ENSG00000284332    0    0             0             0               0   \n",
       "\n",
       "                 Long_ncRNA  Small_ncRNA  Pseudogene  IG_TR_Gene  \n",
       "ENSG00000223972           0            0           1           0  \n",
       "ENSG00000227232           0            0           1           0  \n",
       "ENSG00000278267           0            1           0           0  \n",
       "ENSG00000243485           1            0           0           0  \n",
       "ENSG00000284332           0            1           0           0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add GENCODE biotypes\n",
    "with open(gene_type_conf_path) as gene_type_conf_file:\n",
    "    biotype_dict = yaml.safe_load(gene_type_conf_file)\n",
    "\n",
    "for biotype_category in biotype_dict:\n",
    "    biotype_set = set(biotype_dict[biotype_category])\n",
    "    gene_mat_val_dict = {gene_id: 1 if gene_to_attr_dict[gene_id]['gene_type'] in biotype_set else 0 for gene_id in gene_to_attr_dict}\n",
    "    gene_mat_df[biotype_category] = pd.Series(gene_mat_val_dict)\n",
    "    \n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Update new gene lists\n",
    "The purpose of the following steps is to update new datasets to our gene matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 01. 102 ASD genes (Satterstrom et al., Cell, 2020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>FMRP_targets</th>\n",
       "      <th>Protein_Coding</th>\n",
       "      <th>Long_ncRNA</th>\n",
       "      <th>Small_ncRNA</th>\n",
       "      <th>Pseudogene</th>\n",
       "      <th>IG_TR_Gene</th>\n",
       "      <th>ASD102</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  FMRP_targets  Protein_Coding  \\\n",
       "ENSG00000223972    0    0             0             0               0   \n",
       "ENSG00000227232    0    0             0             0               0   \n",
       "ENSG00000278267    0    0             0             0               0   \n",
       "ENSG00000243485    0    0             0             0               0   \n",
       "ENSG00000284332    0    0             0             0               0   \n",
       "\n",
       "                 Long_ncRNA  Small_ncRNA  Pseudogene  IG_TR_Gene  ASD102  \n",
       "ENSG00000223972           0            0           1           0       0  \n",
       "ENSG00000227232           0            0           1           0       0  \n",
       "ENSG00000278267           0            1           0           0       0  \n",
       "ENSG00000243485           1            0           0           0       0  \n",
       "ENSG00000284332           0            1           0           0       0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "asd_gene_list_path = os.path.join(project_dir, path_dict['ASD'])\n",
    "asd_df = pd.read_excel(asd_gene_list_path, sheet_name='102_ASD')\n",
    "asd_df.dropna(inplace=True)\n",
    "\n",
    "# Find the 102 ASD genes from the genes of the gene matrix\n",
    "asd_gene_set = set(asd_df['ensembl_gene_id'].values)\n",
    "is_asd_gene = np.vectorize(lambda gene_id: 1 if gene_id in asd_gene_set else 0)(gene_mat_df.index.values)\n",
    "\n",
    "# Update the gene list to the gene matrix\n",
    "asd_colname = gene_list_name_dict['ASD']\n",
    "gene_mat_df[asd_colname] = is_asd_gene\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 02. 299 DDD genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>FMRP_targets</th>\n",
       "      <th>Protein_Coding</th>\n",
       "      <th>Long_ncRNA</th>\n",
       "      <th>Small_ncRNA</th>\n",
       "      <th>Pseudogene</th>\n",
       "      <th>IG_TR_Gene</th>\n",
       "      <th>ASD102</th>\n",
       "      <th>DDD299</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  FMRP_targets  Protein_Coding  \\\n",
       "ENSG00000223972    0    0             0             0               0   \n",
       "ENSG00000227232    0    0             0             0               0   \n",
       "ENSG00000278267    0    0             0             0               0   \n",
       "ENSG00000243485    0    0             0             0               0   \n",
       "ENSG00000284332    0    0             0             0               0   \n",
       "\n",
       "                 Long_ncRNA  Small_ncRNA  Pseudogene  IG_TR_Gene  ASD102  \\\n",
       "ENSG00000223972           0            0           1           0       0   \n",
       "ENSG00000227232           0            0           1           0       0   \n",
       "ENSG00000278267           0            1           0           0       0   \n",
       "ENSG00000243485           1            0           0           0       0   \n",
       "ENSG00000284332           0            1           0           0       0   \n",
       "\n",
       "                 DDD299  \n",
       "ENSG00000223972       0  \n",
       "ENSG00000227232       0  \n",
       "ENSG00000278267       0  \n",
       "ENSG00000243485       0  \n",
       "ENSG00000284332       0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "ddd_gene_list_path = os.path.join(project_dir, path_dict['DDD'])\n",
    "ddd_df = pd.read_excel(ddd_gene_list_path, sheet_name='kaplanis_samocha_denovoWEST_res')\n",
    "ddd_df = ddd_df[ddd_df['significant'] == True]  # Leave only significant genes\n",
    "ddd_df = ddd_df.astype({'hgnc_id': 'int64'})\n",
    "\n",
    "# FInd the 299 DDD genes from the genes of the gene matrix\n",
    "ddd_hgnc_ids = np.vectorize(lambda x: f'HGNC:{x}')(ddd_df['hgnc_id'].values)\n",
    "ddd_hgnc_id_set = set(ddd_hgnc_ids)\n",
    "gene_id_to_hgnc_id = {gene_id: gene_to_attr_dict[gene_id].get('hgnc_id') for gene_id in gene_to_attr_dict}  \n",
    "is_ddd_gene = np.vectorize(lambda gene_id: 1 if gene_id_to_hgnc_id[gene_id] in ddd_hgnc_id_set else 0)(gene_mat_df.index.values)\n",
    "\n",
    "# Update the gene list to the gene matrix\n",
    "ddd_colname = gene_list_name_dict['DDD']\n",
    "gene_mat_df[ddd_colname] = is_ddd_gene\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 03. Haploinsufficient genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>FMRP_targets</th>\n",
       "      <th>Protein_Coding</th>\n",
       "      <th>Long_ncRNA</th>\n",
       "      <th>Small_ncRNA</th>\n",
       "      <th>Pseudogene</th>\n",
       "      <th>IG_TR_Gene</th>\n",
       "      <th>ASD102</th>\n",
       "      <th>DDD299</th>\n",
       "      <th>HI_GENE_pLI90_gnomAD</th>\n",
       "      <th>HI_GENE_pLI995_gnomAD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  FMRP_targets  Protein_Coding  \\\n",
       "ENSG00000223972    0    0             0             0               0   \n",
       "ENSG00000227232    0    0             0             0               0   \n",
       "ENSG00000278267    0    0             0             0               0   \n",
       "ENSG00000243485    0    0             0             0               0   \n",
       "ENSG00000284332    0    0             0             0               0   \n",
       "\n",
       "                 Long_ncRNA  Small_ncRNA  Pseudogene  IG_TR_Gene  ASD102  \\\n",
       "ENSG00000223972           0            0           1           0       0   \n",
       "ENSG00000227232           0            0           1           0       0   \n",
       "ENSG00000278267           0            1           0           0       0   \n",
       "ENSG00000243485           1            0           0           0       0   \n",
       "ENSG00000284332           0            1           0           0       0   \n",
       "\n",
       "                 DDD299  HI_GENE_pLI90_gnomAD  HI_GENE_pLI995_gnomAD  \n",
       "ENSG00000223972       0                     0                      0  \n",
       "ENSG00000227232       0                     0                      0  \n",
       "ENSG00000278267       0                     0                      0  \n",
       "ENSG00000243485       0                     0                      0  \n",
       "ENSG00000284332       0                     0                      0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the dataset\n",
    "gnomad_gene_list_path = os.path.join(project_dir, path_dict['GNOMAD_GENE'])\n",
    "gnomad_gene_df = pd.read_table(gnomad_gene_list_path)\n",
    "\n",
    "# Extract HI genes (HI: Haploinsufficient)\n",
    "is_hi_gene_func = lambda pli: pli is not np.nan and pli >= 0.9\n",
    "is_hi_gene = np.vectorize(is_hi_gene_func)(gnomad_gene_df['pLI'].values)\n",
    "hi_gene_df = gnomad_gene_df[is_hi_gene]  \n",
    "\n",
    "# Dictionaries for pLI scores\n",
    "tx_to_pli = {}\n",
    "symbol_to_pli = {}\n",
    "gene_to_pli = {}\n",
    "\n",
    "# Update the 'tx_to_pli' and the 'symbol_to_pli'\n",
    "hi_txs = hi_gene_df['transcript'].values\n",
    "hi_symbols = hi_gene_df['gene'].values\n",
    "hi_plis = hi_gene_df['pLI'].values\n",
    "\n",
    "for tx_id, symbol, pli_score in zip(hi_txs, hi_symbols, hi_plis):\n",
    "    tx_to_pli[tx_id] = pli_score\n",
    "    prev_symbol_pli = symbol_to_pli.get(symbol)\n",
    "    \n",
    "    # Choose the maximum pLI score for duplicated symbols.\n",
    "    if prev_symbol_pli is None or prev_symbol_pli < pli_score:  \n",
    "        symbol_to_pli[symbol] = pli_score\n",
    "\n",
    "# There are 3 steps to update the 'gene_to_pli'\n",
    "# Step 1: Update the 'gene_to_pli' dictionary by getting gene IDs of the transcript IDs via the GENCODE\n",
    "depr_tx_set = set()\n",
    "hi_tx_list = hi_gene_df['transcript'].values\n",
    "\n",
    "for tx_id in hi_tx_list:\n",
    "    attr_dict = tx_to_attr_dict.get(tx_id)\n",
    "    \n",
    "    if attr_dict is None:\n",
    "        depr_tx_set.add(tx_id)\n",
    "    else:\n",
    "        gene_id = attr_dict['Parent'].split('.')[0]\n",
    "        prev_gene_pli = gene_to_pli.get(gene_id)\n",
    "        pli_score = tx_to_pli[tx_id]\n",
    "        \n",
    "        # Choose the maximum pLI score for duplicated genes.\n",
    "        if prev_gene_pli is None or prev_gene_pli < pli_score:    \n",
    "            gene_to_pli[gene_id] = pli_score\n",
    "\n",
    "# Step 2: Update the 'gene_to_pli' dictionary by getting gene IDs of the deprecated transcript IDs using \n",
    "depr_hi_gene_df = hi_gene_df[np.vectorize(lambda tx_id: tx_id in depr_tx_set)(hi_txs)]\n",
    "depr_gene_symbols = depr_hi_gene_df['gene'].values\n",
    "\n",
    "for depr_gene_symbol in depr_gene_symbols:\n",
    "    gene_ids = find_gene_ids(depr_gene_symbol)\n",
    "    \n",
    "    if gene_ids is None:\n",
    "        print(f'{depr_gene_symbol} cannot be found in both the GENCODE and the HGNC.')\n",
    "        continue\n",
    "    \n",
    "    # Update only if the same gene ID does not exist\n",
    "    for gene_id in gene_ids:\n",
    "        if gene_to_pli.get(gene_id) is None:  \n",
    "            gene_to_pli[gene_id] = symbol_to_pli[depr_gene_symbol]\n",
    "\n",
    "# Update the gene matrix\n",
    "# HC: High-confident\n",
    "is_hi_gene = np.vectorize(lambda gene_id: 0 if gene_to_pli.get(gene_id) is None else 1)(gene_mat_df.index.values)  # pLI score >= 0.9\n",
    "is_hc_hi_gene = np.vectorize(lambda gene_id: 1 if gene_to_pli.get(gene_id) is not None and gene_to_pli.get(gene_id) >= 0.995 else 0)(gene_mat_df.index.values)  # pLI score >= 0.995\n",
    "gene_mat_df[gene_list_name_dict['GNOMAD_PLI90']] = is_hi_gene\n",
    "gene_mat_df[gene_list_name_dict['GNOMAD_PLI995']] = is_hc_hi_gene\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 04. Co-expression network analysis modules associated with ASD (Parikshak et al., Nature, 2016) \n",
    "- Genome wide transcriptome analysis of 251 post-mortem samples of frontal and temporal cortex and cerebellum from 48 individuals with ASD and 49 control subjects\n",
    "- Identified 6 modules significantly associated with ASD of 24 modules from WGCNA in the cortex analysis\n",
    "- Upregulated: CTX.M9, CTX.M19, CTX.M20\n",
    "- Downregulated: CTX.M4, CTX.M10, CTX.M16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASD_WGCNA_Module_CTX.M4 (Total no.genes: 250)\n",
      "ENSG00000235477 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000244567 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000226852 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000261496 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000266753 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000272327 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "ASD_WGCNA_Module_CTX.M9 (Total no.genes: 525)\n",
      "ENSG00000241732 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000261662 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "ASD_WGCNA_Module_CTX.M10 (Total no.genes: 259)\n",
      "\n",
      "ASD_WGCNA_Module_CTX.M16 (Total no.genes: 284)\n",
      "ENSG00000260822 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000260415 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000260432 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000260837 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "ASD_WGCNA_Module_CTX.M19 (Total no.genes: 274)\n",
      "\n",
      "ASD_WGCNA_Module_CTX.M20 (Total no.genes: 348)\n",
      "ENSG00000224849 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000269728 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>...</th>\n",
       "      <th>ASD102</th>\n",
       "      <th>DDD299</th>\n",
       "      <th>HI_GENE_pLI90_gnomAD</th>\n",
       "      <th>HI_GENE_pLI995_gnomAD</th>\n",
       "      <th>ASD_WGCNA_Module_CTX.M4</th>\n",
       "      <th>ASD_WGCNA_Module_CTX.M9</th>\n",
       "      <th>ASD_WGCNA_Module_CTX.M10</th>\n",
       "      <th>ASD_WGCNA_Module_CTX.M16</th>\n",
       "      <th>ASD_WGCNA_Module_CTX.M19</th>\n",
       "      <th>ASD_WGCNA_Module_CTX.M20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  ...  ASD102  DDD299  \\\n",
       "ENSG00000223972    0    0             0  ...       0       0   \n",
       "ENSG00000227232    0    0             0  ...       0       0   \n",
       "ENSG00000278267    0    0             0  ...       0       0   \n",
       "ENSG00000243485    0    0             0  ...       0       0   \n",
       "ENSG00000284332    0    0             0  ...       0       0   \n",
       "\n",
       "                 HI_GENE_pLI90_gnomAD  HI_GENE_pLI995_gnomAD  \\\n",
       "ENSG00000223972                     0                      0   \n",
       "ENSG00000227232                     0                      0   \n",
       "ENSG00000278267                     0                      0   \n",
       "ENSG00000243485                     0                      0   \n",
       "ENSG00000284332                     0                      0   \n",
       "\n",
       "                 ASD_WGCNA_Module_CTX.M4  ASD_WGCNA_Module_CTX.M9  \\\n",
       "ENSG00000223972                        0                        0   \n",
       "ENSG00000227232                        0                        0   \n",
       "ENSG00000278267                        0                        0   \n",
       "ENSG00000243485                        0                        0   \n",
       "ENSG00000284332                        0                        0   \n",
       "\n",
       "                 ASD_WGCNA_Module_CTX.M10  ASD_WGCNA_Module_CTX.M16  \\\n",
       "ENSG00000223972                         0                         0   \n",
       "ENSG00000227232                         0                         0   \n",
       "ENSG00000278267                         0                         0   \n",
       "ENSG00000243485                         0                         0   \n",
       "ENSG00000284332                         0                         0   \n",
       "\n",
       "                 ASD_WGCNA_Module_CTX.M19  ASD_WGCNA_Module_CTX.M20  \n",
       "ENSG00000223972                         0                         0  \n",
       "ENSG00000227232                         0                         0  \n",
       "ENSG00000278267                         0                         0  \n",
       "ENSG00000243485                         0                         0  \n",
       "ENSG00000284332                         0                         0  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "mat_col_prefix = gene_list_name_dict['ASD_GWTA']\n",
    "parik_table_path = os.path.join(project_dir, path_dict['ASD_GWTA'])\n",
    "parik_gene_df = pd.read_excel(parik_table_path, sheet_name='TableS2a', header=1)\n",
    "\n",
    "# Split the genes by their modules\n",
    "groupby_module = parik_gene_df.groupby('WGCNA Module Label')\n",
    "asd_module_nums = [4, 9, 10, 16, 19, 20]  # Modules significantly associated with ASD\n",
    "asd_module_dfs = [groupby_module.get_group(asd_module_num) for asd_module_num in asd_module_nums]\n",
    "\n",
    "# Update the gene matrix\n",
    "gencode_gene_ids = gene_mat_df.index.values\n",
    "gencode_gene_id_set = set(gencode_gene_ids)\n",
    "\n",
    "for i, asd_module_df in enumerate(asd_module_dfs):\n",
    "    module_num = asd_module_nums[i]\n",
    "    module_name = f'CTX.M{module_num}'\n",
    "    mat_col_name = f\"{mat_col_prefix}_{module_name}\"\n",
    "    mod_gene_cnt = len(asd_module_df.index.values)\n",
    "    print(f'{mat_col_name} (Total no.genes: {mod_gene_cnt})')\n",
    "    \n",
    "    # Make a set of gene IDs in the module\n",
    "    mod_gene_ids = asd_module_df['ENSEMBL ID'].values\n",
    "    mod_gene_id_set = set()\n",
    "    mod_gene_symbols = asd_module_df['HGNC Symbol'].values\n",
    "    \n",
    "    for i in range(mod_gene_cnt):\n",
    "        mod_gene_id = mod_gene_ids[i]\n",
    "        mod_gene_symbol = mod_gene_symbols[i]\n",
    "        \n",
    "        if mod_gene_id in gencode_gene_id_set:\n",
    "            mod_gene_id_set.add(mod_gene_id)\n",
    "        else:  # Replace with alternative gene IDs\n",
    "            if mod_gene_symbol is np.nan:  # New mod_gene ID cannot be found.\n",
    "                print(f'{mod_gene_id} cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.')\n",
    "            else:\n",
    "                alt_mod_gene_ids = find_gene_ids(mod_gene_symbol)\n",
    "                \n",
    "                for alt_mod_gene_id in alt_mod_gene_ids:\n",
    "                    mod_gene_id_set.add(alt_mod_gene_id)\n",
    "    \n",
    "    # Update\n",
    "    gene_mat_vals = np.vectorize(lambda gene_id: 1 if gene_id in mod_gene_id_set else 0)(gencode_gene_ids)\n",
    "    gene_mat_df[mat_col_name] = gene_mat_vals\n",
    "    print()\n",
    "\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 05. Nowakowski et al., Science, 2017\n",
    "- scRNA-seq for primary cortical and medial ganglionic eminence (MGE) in developing human telencephalon\n",
    "- From 48 samples with 5.85 ~ 37 PCW\n",
    "- Unbiased clustering and found marker genes for each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now2017_SC_Marker_Astrocyte (Total no.genes: 976)\n",
      "38 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_Choroid (Total no.genes: 1493)\n",
      "90 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_EN-PFC1 (Total no.genes: 1890)\n",
      "135 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_EN-PFC2 (Total no.genes: 1896)\n",
      "86 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_EN-PFC3 (Total no.genes: 556)\n",
      "41 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_EN-V1-1 (Total no.genes: 1693)\n",
      "85 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_EN-V1-2 (Total no.genes: 1394)\n",
      "70 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_EN-V1-3 (Total no.genes: 1841)\n",
      "105 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_Endothelial (Total no.genes: 2841)\n",
      "123 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_Glyc (Total no.genes: 1524)\n",
      "85 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IN-CTX-CGE1 (Total no.genes: 1324)\n",
      "51 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IN-CTX-CGE2 (Total no.genes: 1366)\n",
      "64 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IN-CTX-MGE1 (Total no.genes: 1257)\n",
      "50 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IN-CTX-MGE2 (Total no.genes: 1300)\n",
      "82 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IN-STR (Total no.genes: 1162)\n",
      "50 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IPC-div1 (Total no.genes: 1037)\n",
      "41 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IPC-div2 (Total no.genes: 1409)\n",
      "48 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IPC-nEN1 (Total no.genes: 476)\n",
      "20 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IPC-nEN2 (Total no.genes: 787)\n",
      "30 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_IPC-nEN3 (Total no.genes: 482)\n",
      "42 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_MGE-IPC1 (Total no.genes: 1068)\n",
      "40 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_MGE-IPC2 (Total no.genes: 991)\n",
      "34 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_MGE-IPC3 (Total no.genes: 645)\n",
      "30 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_MGE-RG1 (Total no.genes: 750)\n",
      "26 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_MGE-RG2 (Total no.genes: 1088)\n",
      "44 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_MGE-div (Total no.genes: 1254)\n",
      "51 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_Microglia (Total no.genes: 2804)\n",
      "141 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_Mural (Total no.genes: 1356)\n",
      "57 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_OPC (Total no.genes: 647)\n",
      "22 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_RG-div1 (Total no.genes: 3304)\n",
      "114 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_RG-div2 (Total no.genes: 2410)\n",
      "66 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_RG-early (Total no.genes: 1309)\n",
      "67 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_U1 (Total no.genes: 2552)\n",
      "111 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_U2 (Total no.genes: 3141)\n",
      "118 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_U3 (Total no.genes: 7026)\n",
      "428 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_U4 (Total no.genes: 4185)\n",
      "260 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nEN-early1 (Total no.genes: 1226)\n",
      "65 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nEN-early2 (Total no.genes: 1007)\n",
      "48 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nEN-late (Total no.genes: 1150)\n",
      "38 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nIN1 (Total no.genes: 934)\n",
      "36 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nIN2 (Total no.genes: 2128)\n",
      "74 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nIN3 (Total no.genes: 870)\n",
      "39 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nIN4 (Total no.genes: 840)\n",
      "38 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_nIN5 (Total no.genes: 1438)\n",
      "51 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_oRG (Total no.genes: 2554)\n",
      "114 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_tRG (Total no.genes: 1918)\n",
      "71 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Now2017_SC_Marker_vRG (Total no.genes: 915)\n",
      "36 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>...</th>\n",
       "      <th>Now2017_SC_Marker_nEN-early2</th>\n",
       "      <th>Now2017_SC_Marker_nEN-late</th>\n",
       "      <th>Now2017_SC_Marker_nIN1</th>\n",
       "      <th>Now2017_SC_Marker_nIN2</th>\n",
       "      <th>Now2017_SC_Marker_nIN3</th>\n",
       "      <th>Now2017_SC_Marker_nIN4</th>\n",
       "      <th>Now2017_SC_Marker_nIN5</th>\n",
       "      <th>Now2017_SC_Marker_oRG</th>\n",
       "      <th>Now2017_SC_Marker_tRG</th>\n",
       "      <th>Now2017_SC_Marker_vRG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  ...  Now2017_SC_Marker_nEN-early2  \\\n",
       "ENSG00000223972    0    0             0  ...                             0   \n",
       "ENSG00000227232    0    0             0  ...                             0   \n",
       "ENSG00000278267    0    0             0  ...                             0   \n",
       "ENSG00000243485    0    0             0  ...                             0   \n",
       "ENSG00000284332    0    0             0  ...                             0   \n",
       "\n",
       "                 Now2017_SC_Marker_nEN-late  Now2017_SC_Marker_nIN1  \\\n",
       "ENSG00000223972                           0                       0   \n",
       "ENSG00000227232                           0                       0   \n",
       "ENSG00000278267                           0                       0   \n",
       "ENSG00000243485                           0                       0   \n",
       "ENSG00000284332                           0                       0   \n",
       "\n",
       "                 Now2017_SC_Marker_nIN2  Now2017_SC_Marker_nIN3  \\\n",
       "ENSG00000223972                       0                       0   \n",
       "ENSG00000227232                       0                       0   \n",
       "ENSG00000278267                       0                       0   \n",
       "ENSG00000243485                       0                       0   \n",
       "ENSG00000284332                       0                       0   \n",
       "\n",
       "                 Now2017_SC_Marker_nIN4  Now2017_SC_Marker_nIN5  \\\n",
       "ENSG00000223972                       0                       0   \n",
       "ENSG00000227232                       0                       0   \n",
       "ENSG00000278267                       0                       0   \n",
       "ENSG00000243485                       0                       0   \n",
       "ENSG00000284332                       0                       0   \n",
       "\n",
       "                 Now2017_SC_Marker_oRG  Now2017_SC_Marker_tRG  \\\n",
       "ENSG00000223972                      0                      0   \n",
       "ENSG00000227232                      0                      0   \n",
       "ENSG00000278267                      0                      0   \n",
       "ENSG00000243485                      0                      0   \n",
       "ENSG00000284332                      0                      0   \n",
       "\n",
       "                 Now2017_SC_Marker_vRG  \n",
       "ENSG00000223972                      0  \n",
       "ENSG00000227232                      0  \n",
       "ENSG00000278267                      0  \n",
       "ENSG00000243485                      0  \n",
       "ENSG00000284332                      0  \n",
       "\n",
       "[5 rows x 73 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "now_table_path = os.path.join(project_dir, path_dict['SC_Now2017'])\n",
    "now_gene_df = pd.read_excel(now_table_path, sheet_name='Table5 - Clustermarkers')\n",
    "\n",
    "# Split the genes by their modules\n",
    "mat_col_prefix = gene_list_name_dict['SC_Now2017']\n",
    "groupby_cluster = now_gene_df.groupby('cluster')\n",
    "cluster_names = list(groupby_cluster.groups.keys())\n",
    "gene_cluster_dfs = [groupby_cluster.get_group(cluster_name) for cluster_name in cluster_names]\n",
    "\n",
    "# Update the gene matrix\n",
    "mat_indices = gene_mat_df.index.values\n",
    "mat_index_set = set(mat_indices)\n",
    "\n",
    "for i, gene_cluster_df in enumerate(gene_cluster_dfs):\n",
    "    cluster_name = cluster_names[i]\n",
    "    mat_col_name = f'{mat_col_prefix}_{cluster_name}'\n",
    "    marker_genes = gene_cluster_df['gene'].values\n",
    "    marker_gene_cnt = len(marker_genes)\n",
    "    print(f'{mat_col_name} (Total no.genes: {marker_gene_cnt})')\n",
    "    \n",
    "    marker_gene_id_set = set()\n",
    "    na_gene_cnt = 0\n",
    "    \n",
    "    # Update\n",
    "    for marker_gene in marker_genes:\n",
    "        marker_gene_ids = find_gene_ids(marker_gene)\n",
    "        \n",
    "        if marker_gene_ids is None:\n",
    "            na_gene_cnt += 1\n",
    "        else:\n",
    "            for marker_gene_id in marker_gene_ids:\n",
    "                marker_gene_id_set.add(marker_gene_id)\n",
    "            \n",
    "    gene_mat_vals = np.vectorize(lambda gene_id: 1 if gene_id in marker_gene_id_set else 0)(mat_indices)\n",
    "    gene_mat_df[mat_col_name] = gene_mat_vals \n",
    "    \n",
    "    print(f'{na_gene_cnt} genes cannot be found in both the GENCODE v33 and the HGNC.')\n",
    "    print()\n",
    "\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 06. Li et al., Science, 2018\n",
    "- Single-cell from 9 de-identified postmortem brains (Table S3)\n",
    "- 5 PCW ~ 20 PCW\n",
    "- Marker genes for each cell type of Pallium, CP, DFC, NCX regions in a human brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Li2018_SC_Marker_Astro_adult (Total no.genes: 46)\n",
      "0 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Astro_prenatal (Total no.genes: 37)\n",
      "3 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Endo_adult (Total no.genes: 39)\n",
      "1 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Endo_prenatal (Total no.genes: 178)\n",
      "10 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_ExN_adult (Total no.genes: 27)\n",
      "0 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_ExN_prenatal (Total no.genes: 13)\n",
      "2 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_IntN_adult (Total no.genes: 22)\n",
      "2 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_IntN_prenatal (Total no.genes: 33)\n",
      "6 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Microglia_adult (Total no.genes: 40)\n",
      "1 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Microglia_prenatal (Total no.genes: 197)\n",
      "5 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_NPC_prenatal (Total no.genes: 40)\n",
      "2 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_OPC_adult (Total no.genes: 33)\n",
      "0 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Oligo_adult (Total no.genes: 43)\n",
      "0 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_Oligo_prenatal (Total no.genes: 33)\n",
      "3 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n",
      "Li2018_SC_Marker_VSMC_adult (Total no.genes: 45)\n",
      "0 genes cannot be found in both the GENCODE v33 and the HGNC.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>...</th>\n",
       "      <th>Li2018_SC_Marker_ExN_prenatal</th>\n",
       "      <th>Li2018_SC_Marker_IntN_adult</th>\n",
       "      <th>Li2018_SC_Marker_IntN_prenatal</th>\n",
       "      <th>Li2018_SC_Marker_Microglia_adult</th>\n",
       "      <th>Li2018_SC_Marker_Microglia_prenatal</th>\n",
       "      <th>Li2018_SC_Marker_NPC_prenatal</th>\n",
       "      <th>Li2018_SC_Marker_OPC_adult</th>\n",
       "      <th>Li2018_SC_Marker_Oligo_adult</th>\n",
       "      <th>Li2018_SC_Marker_Oligo_prenatal</th>\n",
       "      <th>Li2018_SC_Marker_VSMC_adult</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  ...  Li2018_SC_Marker_ExN_prenatal  \\\n",
       "ENSG00000223972    0    0             0  ...                              0   \n",
       "ENSG00000227232    0    0             0  ...                              0   \n",
       "ENSG00000278267    0    0             0  ...                              0   \n",
       "ENSG00000243485    0    0             0  ...                              0   \n",
       "ENSG00000284332    0    0             0  ...                              0   \n",
       "\n",
       "                 Li2018_SC_Marker_IntN_adult  Li2018_SC_Marker_IntN_prenatal  \\\n",
       "ENSG00000223972                            0                               0   \n",
       "ENSG00000227232                            0                               0   \n",
       "ENSG00000278267                            0                               0   \n",
       "ENSG00000243485                            0                               0   \n",
       "ENSG00000284332                            0                               0   \n",
       "\n",
       "                 Li2018_SC_Marker_Microglia_adult  \\\n",
       "ENSG00000223972                                 0   \n",
       "ENSG00000227232                                 0   \n",
       "ENSG00000278267                                 0   \n",
       "ENSG00000243485                                 0   \n",
       "ENSG00000284332                                 0   \n",
       "\n",
       "                 Li2018_SC_Marker_Microglia_prenatal  \\\n",
       "ENSG00000223972                                    0   \n",
       "ENSG00000227232                                    0   \n",
       "ENSG00000278267                                    0   \n",
       "ENSG00000243485                                    0   \n",
       "ENSG00000284332                                    0   \n",
       "\n",
       "                 Li2018_SC_Marker_NPC_prenatal  Li2018_SC_Marker_OPC_adult  \\\n",
       "ENSG00000223972                              0                           0   \n",
       "ENSG00000227232                              0                           0   \n",
       "ENSG00000278267                              0                           0   \n",
       "ENSG00000243485                              0                           0   \n",
       "ENSG00000284332                              0                           0   \n",
       "\n",
       "                 Li2018_SC_Marker_Oligo_adult  \\\n",
       "ENSG00000223972                             0   \n",
       "ENSG00000227232                             0   \n",
       "ENSG00000278267                             0   \n",
       "ENSG00000243485                             0   \n",
       "ENSG00000284332                             0   \n",
       "\n",
       "                 Li2018_SC_Marker_Oligo_prenatal  Li2018_SC_Marker_VSMC_adult  \n",
       "ENSG00000223972                                0                            0  \n",
       "ENSG00000227232                                0                            0  \n",
       "ENSG00000278267                                0                            0  \n",
       "ENSG00000243485                                0                            0  \n",
       "ENSG00000284332                                0                            0  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "li_table_path = os.path.join(project_dir, path_dict['SC_Li2018'])\n",
    "li_gene_df = pd.read_excel(li_table_path, sheet_name='Table S8', header=3)\n",
    "\n",
    "# Split the dataframe into two DataFrames and concatenate them\n",
    "prenatal_gene_df = li_gene_df[['Gene symbol', 'Cell type', 'Age']].copy()\n",
    "adult_gene_df = li_gene_df[['Gene symbol.1', 'Cell type.1', 'Age.1']].copy()\n",
    "adult_gene_df.rename(columns={colname: colname.split('.')[0] for colname in adult_gene_df.columns.values}, inplace=True)\n",
    "adult_gene_df.dropna(inplace=True)\n",
    "li_gene_df = pd.concat([prenatal_gene_df, adult_gene_df], ignore_index=True)\n",
    "\n",
    "# Split the genes by their cell types\n",
    "mat_col_prefix = gene_list_name_dict['SC_Li2018']\n",
    "groupby_cell = li_gene_df.groupby('Cell type')\n",
    "cell_names = list(groupby_cell.groups.keys())\n",
    "cell_gene_dfs = [groupby_cell.get_group(cell_name) for cell_name in cell_names]\n",
    "\n",
    "# Update the gene matrix\n",
    "mat_indices = gene_mat_df.index.values\n",
    "\n",
    "for i, cell_gene_df in enumerate(cell_gene_dfs):\n",
    "    cell_name = cell_names[i]\n",
    "    mat_col_name = f'{mat_col_prefix}_{cell_name}'\n",
    "    marker_genes = cell_gene_df['Gene symbol'].values\n",
    "    marker_gene_cnt = len(marker_genes)\n",
    "    print(f'{mat_col_name} (Total no.genes: {marker_gene_cnt})')\n",
    "    \n",
    "    # Update\n",
    "    marker_gene_id_set = set()\n",
    "    na_gene_cnt = 0\n",
    "    \n",
    "    # Update\n",
    "    for marker_gene in marker_genes:\n",
    "        marker_gene_ids = find_gene_ids(marker_gene)\n",
    "        \n",
    "        if marker_gene_ids is None:\n",
    "            na_gene_cnt += 1\n",
    "        else:\n",
    "            for marker_gene_id in marker_gene_ids:\n",
    "                marker_gene_id_set.add(marker_gene_id)\n",
    "            \n",
    "    gene_mat_vals = np.vectorize(lambda gene_id: 1 if gene_id in marker_gene_id_set else 0)(mat_indices)\n",
    "    gene_mat_df[mat_col_name] = gene_mat_vals \n",
    "    \n",
    "    print(f'{na_gene_cnt} genes cannot be found in both the GENCODE v33 and the HGNC.')\n",
    "    print()\n",
    "\n",
    "gene_mat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 07. Velmeshev et al., Science, 2019\n",
    "- Single-nucleus sequencing \n",
    "- 41 post-mortem tissue samples including prefrontal cortex (PFC) and anterior cingulate cortex (ACC) from 16 controls and 15 ASD patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vel2019_SN_Marker_AST-FB (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_AST-PP (Total no.genes: 45)\n",
      "ENSG00000255864 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "Vel2019_SN_Marker_Endothelial (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_IN-PV (Total no.genes: 45)\n",
      "ENSG00000256248 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000255864 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "Vel2019_SN_Marker_IN-SST (Total no.genes: 45)\n",
      "ENSG00000256248 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000228214 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "Vel2019_SN_Marker_IN-SV2C (Total no.genes: 45)\n",
      "ENSG00000256248 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "ENSG00000228214 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "Vel2019_SN_Marker_IN-VIP (Total no.genes: 45)\n",
      "ENSG00000256248 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "Vel2019_SN_Marker_L2/3 (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_L4 (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_L5/6 (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_L5/6-CC (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_Microglia (Total no.genes: 45)\n",
      "ENSG00000228651 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n",
      "Vel2019_SN_Marker_Neu-NRGN-I (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_Neu-NRGN-II (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_Neu-mat (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_OPC (Total no.genes: 45)\n",
      "\n",
      "Vel2019_SN_Marker_Oligodendrocytes (Total no.genes: 45)\n",
      "ENSG00000243444 cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ASD_TADA_FDR01</th>\n",
       "      <th>ASD_TADA_FDR03</th>\n",
       "      <th>ASD_midfetal_coexpression</th>\n",
       "      <th>BE</th>\n",
       "      <th>PLI90Score</th>\n",
       "      <th>PSD</th>\n",
       "      <th>DDD</th>\n",
       "      <th>CHD8_targets</th>\n",
       "      <th>...</th>\n",
       "      <th>Vel2019_SN_Marker_L2/3</th>\n",
       "      <th>Vel2019_SN_Marker_L4</th>\n",
       "      <th>Vel2019_SN_Marker_L5/6</th>\n",
       "      <th>Vel2019_SN_Marker_L5/6-CC</th>\n",
       "      <th>Vel2019_SN_Marker_Microglia</th>\n",
       "      <th>Vel2019_SN_Marker_Neu-NRGN-I</th>\n",
       "      <th>Vel2019_SN_Marker_Neu-NRGN-II</th>\n",
       "      <th>Vel2019_SN_Marker_Neu-mat</th>\n",
       "      <th>Vel2019_SN_Marker_OPC</th>\n",
       "      <th>Vel2019_SN_Marker_Oligodendrocytes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id    gene_name  ASD_TADA_FDR01  \\\n",
       "ENSG00000223972  ENSG00000223972.5      DDX11L1               0   \n",
       "ENSG00000227232  ENSG00000227232.5       WASH7P               0   \n",
       "ENSG00000278267  ENSG00000278267.1    MIR6859-1               0   \n",
       "ENSG00000243485  ENSG00000243485.5  MIR1302-2HG               0   \n",
       "ENSG00000284332  ENSG00000284332.1    MIR1302-2               0   \n",
       "\n",
       "                 ASD_TADA_FDR03  ASD_midfetal_coexpression  BE  PLI90Score  \\\n",
       "ENSG00000223972               0                          0   0           0   \n",
       "ENSG00000227232               0                          0   0           0   \n",
       "ENSG00000278267               0                          0   0           0   \n",
       "ENSG00000243485               0                          0   0           0   \n",
       "ENSG00000284332               0                          0   0           0   \n",
       "\n",
       "                 PSD  DDD  CHD8_targets  ...  Vel2019_SN_Marker_L2/3  \\\n",
       "ENSG00000223972    0    0             0  ...                       0   \n",
       "ENSG00000227232    0    0             0  ...                       0   \n",
       "ENSG00000278267    0    0             0  ...                       0   \n",
       "ENSG00000243485    0    0             0  ...                       0   \n",
       "ENSG00000284332    0    0             0  ...                       0   \n",
       "\n",
       "                 Vel2019_SN_Marker_L4  Vel2019_SN_Marker_L5/6  \\\n",
       "ENSG00000223972                     0                       0   \n",
       "ENSG00000227232                     0                       0   \n",
       "ENSG00000278267                     0                       0   \n",
       "ENSG00000243485                     0                       0   \n",
       "ENSG00000284332                     0                       0   \n",
       "\n",
       "                 Vel2019_SN_Marker_L5/6-CC  Vel2019_SN_Marker_Microglia  \\\n",
       "ENSG00000223972                          0                            0   \n",
       "ENSG00000227232                          0                            0   \n",
       "ENSG00000278267                          0                            0   \n",
       "ENSG00000243485                          0                            0   \n",
       "ENSG00000284332                          0                            0   \n",
       "\n",
       "                 Vel2019_SN_Marker_Neu-NRGN-I  Vel2019_SN_Marker_Neu-NRGN-II  \\\n",
       "ENSG00000223972                             0                              0   \n",
       "ENSG00000227232                             0                              0   \n",
       "ENSG00000278267                             0                              0   \n",
       "ENSG00000243485                             0                              0   \n",
       "ENSG00000284332                             0                              0   \n",
       "\n",
       "                 Vel2019_SN_Marker_Neu-mat  Vel2019_SN_Marker_OPC  \\\n",
       "ENSG00000223972                          0                      0   \n",
       "ENSG00000227232                          0                      0   \n",
       "ENSG00000278267                          0                      0   \n",
       "ENSG00000243485                          0                      0   \n",
       "ENSG00000284332                          0                      0   \n",
       "\n",
       "                 Vel2019_SN_Marker_Oligodendrocytes  \n",
       "ENSG00000223972                                   0  \n",
       "ENSG00000227232                                   0  \n",
       "ENSG00000278267                                   0  \n",
       "ENSG00000243485                                   0  \n",
       "ENSG00000284332                                   0  \n",
       "\n",
       "[5 rows x 105 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "vel_table_path = os.path.join(project_dir, path_dict['SN_Vel2019'])\n",
    "vel_gene_df = pd.read_excel(vel_table_path, sheet_name='cell type markers')\n",
    "vel_gene_df.head()\n",
    "\n",
    "# Split the genes by their modules\n",
    "mat_col_prefix = gene_list_name_dict['SN_Vel2019']\n",
    "groupby_cell = vel_gene_df.groupby('Cell type')\n",
    "cell_names = list(groupby_cell.groups.keys())\n",
    "cell_gene_dfs = [groupby_cell.get_group(cell_name) for cell_name in cell_names]\n",
    "\n",
    "# Update the gene matrix\n",
    "mat_indices = gene_mat_df.index.values\n",
    "mat_index_set = set(mat_indices)\n",
    "\n",
    "for i, cell_gene_df in enumerate(cell_gene_dfs):\n",
    "    cell_name = cell_names[i]\n",
    "    mat_col_name = f'{mat_col_prefix}_{cell_name}'\n",
    "    marker_gene_ids = cell_gene_df['Gene ID'].values\n",
    "    marker_gene_names = cell_gene_df['Gene name'].values\n",
    "    marker_gene_cnt = len(marker_genes)\n",
    "    print(f'{mat_col_name} (Total no.genes: {marker_gene_cnt})')\n",
    "    \n",
    "    marker_gene_id_set = set()\n",
    "    na_gene_cnt = 0\n",
    "    \n",
    "    for marker_gene_id, marker_gene_name in zip(marker_gene_ids, marker_gene_names):\n",
    "        if marker_gene_id in mat_index_set:\n",
    "            marker_gene_id_set.add(marker_gene_id)\n",
    "        else:\n",
    "            alt_gene_ids = find_gene_ids(marker_gene_name)\n",
    "            \n",
    "            if alt_gene_ids is None:\n",
    "                print(f'{marker_gene_id} cannot be found in both the GENCODE and the HGNC, so this gene will be excluded.')\n",
    "            else:\n",
    "                for alt_gene_id in alt_gene_ids:\n",
    "                    marker_gene_id_set.add(alt_gene_id)\n",
    "                    \n",
    "    gene_mat_vals = np.vectorize(lambda gene_id: 1 if gene_id in marker_gene_id_set else 0)(mat_indices)\n",
    "    gene_mat_df[mat_col_name] = gene_mat_vals            \n",
    "    print()\n",
    "    \n",
    "gene_mat_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
